chemprop/models/multi.py

```diff
@@ -46,3 +46,38 @@ def fingerprint(
         H = self.bn(H)
 
         return H if X_f is None else torch.cat((H, X_f), 1)
+
+    @classmethod
+    def load_from_checkpoint(
+        cls, checkpoint_path, map_location=None, hparams_file=None, strict=True, **kwargs
+    ) -> MPNN:
+        hparams = torch.load(checkpoint_path)["hyper_parameters"]
+
+        kwargs = construct_modules(hparams, **kwargs)
+
+        return super().load_from_checkpoint(
+            checkpoint_path, map_location, hparams_file, strict, **kwargs
+        )
+
+
+def construct_modules(hparams: dict, **kwargs):
+    # construct the `agg` and `predictor` modules from their hparam dicts like before
+    # and store them into `kwargs`
+    agg_hparams = hparams["agg"]
+    agg = agg_hparams.pop("cls")(**agg_hparams)
+    kwargs["agg"] = agg
+
+    predictor_hparams = hparams["predictor"]
+    predictor = predictor_hparams.pop("cls")(**predictor_hparams)
+    kwargs["predictor"] = predictor
+
+    # iteratively construct the input message blocks from their corresponding hparams
+    # we do this because the `MulticomponentMessagePassing.hparams` attribute is a nested
+    # dictionary of hparams 
+    mp_hparams = hparams["message_passing"]
+    mp_hparams["blocks"] = [
+        block_hparams.pop("cls")(**block_hparams) for block_hparams in mp_hparams["blocks"]
+    ]
+    message_passing = mp_hparams.pop("cls")(**mp_hparams)
+    kwargs["message_passing"] = message_passing
+    return kwargs
```

Comment:
can you just move this into the function? it's not called anywhere else in the code.

Reply:
Thanks for the question! Unfortunately, if these lines are moved into the function, it raises `KeyError` similar to the one reported in https://github.com/Lightning-AI/pytorch-lightning/issues/18405. The error traces it raises are shown below. I move this into a standalone function as suggested by this [comment](https://github.com/Lightning-AI/pytorch-lightning/issues/18405#issuecomment-1746330593). I think we might need to stick to this until the `KeyError` is resolved by pytorch-lightning.

```
Exception has occurred: KeyError
'self'
  File "/Users/hwpang/Software/chemprop/chemprop/nn/predictors.py", line 76, in __init__
    self.save_hyperparameters()
  File "/Users/hwpang/Software/chemprop/chemprop/nn/predictors.py", line 121, in __init__
    super().__init__(n_tasks, input_dim, hidden_dim, n_layers, dropout, activation, criterion)
  File "/Users/hwpang/Software/chemprop/chemprop/models/multi.py", line 63, in load_from_checkpoint
    predictor = predictor_hparams.pop("cls")(**predictor_hparams)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/hwpang/Software/chemprop/examples/test.py", line 24, in <module>
    mcmpnn = multi.MulticomponentMPNN.load_from_checkpoint(checkpoint_path)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
KeyError: 'self'
```